src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,152 @@
+            Node<T> newNode = new Node<>(prevNode, value, currentNode);
+            currentNode.prev = newNode;
+            prevNode.next = newNode;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T previousValue = currentNode.element;
+        currentNode.element = value;
+        return previousValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> thisNode = getNodeByValue(object);
+        if (thisNode != null) {
+            unlink(thisNode);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private class Node<T> {
+        private T element;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.element = element;
+            this.next = next;
+            this.prev = prev;
+            size++;
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(INDEX_EXCEPTION);
+        }
+    }
+
+    private T unlink(Node<T> node) {
+        Node<T> nextNode = node.next;
+        Node<T> prevNode = node.prev;
+        if (prevNode == null) {
+            head = nextNode;
+        } else {
+            prevNode.next = nextNode;
+        }
+        if (nextNode == null) {
+            tail = prevNode;
+        } else {
+            nextNode.prev = prevNode;
+        }
+        size--;
+        return node.element;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> currentNode = head;
+        if (index >= size * SIZE_DEGREE) {
+            currentNode = tail;
+            for (int i = index; i < size - 1; i++) {
+                currentNode = currentNode.prev;
+            }
+        } else {
+            for (int i = index; i > 0; i--) {
+                currentNode = currentNode.next;
+            }
+        }
+        return currentNode;
+    }
+
+    private Node<T> getNodeByValue(T value) {
+        for (int i = 0; i < size; i++) {
+            if (getNodeByIndex(i).element == value
+                    || value != null && value.equals(getNodeByIndex(i).element)) {
+                return getNodeByIndex(i);
```

Comment:
it's a quite complicated algorithm. could take some time. let's not use `getNodeByIndex`, lets iterate from head with `node.next`